File: dotnet/Zirve.Sdk/Crm/CrmManager.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Zirve.Sdk.Config;

namespace Zirve.Sdk.Crm;

public class CrmManager
{
    private readonly HttpClient _httpClient;
    private readonly ConfigManager _config;
    private readonly string _url;
    private readonly string _db;
    private int? _uid;
    private int _rpcId = 0;

    public CrmManager(HttpClient httpClient, ConfigManager configManager)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        _config = configManager ?? throw new ArgumentNullException(nameof(configManager));
        
        var cfg = _config.Module("crm");
        _url = (cfg.GetValueOrDefault("url", "http://odoo.zirve-infra.svc.cluster.local:8069") ?? "").TrimEnd('/');
        _db = cfg.GetValueOrDefault("database", "zirve") ?? "zirve";
    }

    private async Task<JsonElement> RpcAsync(string method, object paramsBody)
    {
        _rpcId++;
        
        var payload = new
        {
            jsonrpc = "2.0",
            method = method,
            @params = paramsBody,
            id = _rpcId
        };

        var request = new HttpRequestMessage(HttpMethod.Post, $"{_url}/jsonrpc")
        {
            Content = new StringContent(JsonSerializer.Serialize(payload), System.Text.Encoding.UTF8, "application/json")
        };

        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();

        using var doc = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
        
        if (doc.RootElement.TryGetProperty("error", out var err))
        {
            throw new Exception($"Odoo RPC Error: {err.GetRawText()}");
        }

        return doc.RootElement.GetProperty("result").Clone();
    }

    private async Task<int> AuthenticateAsync()
    {
        if (_uid.HasValue) return _uid.Value;

        var cfg = _config.Module("crm");
        var user = cfg.GetValueOrDefault("username", "admin");
        var password = cfg.GetValueOrDefault("password", "admin");

        var result = await RpcAsync("call", new
        {
            service = "common",
            method = "authenticate",
            args = new object[] { _db, user!, password!, new { } }
        });

        if (result.ValueKind == JsonValueKind.False || result.ValueKind == JsonValueKind.Null)
        {
            throw new Exception("Odoo Authentication failed.");
        }

        _uid = result.GetInt32();
        return _uid.Value;
    }

    private async Task<JsonElement> ExecuteKwAsync(string model, string method, object[] args, Dictionary<string, object>? kwargs = null)
    {
        var uid = await AuthenticateAsync();
        var password = _config.Module("crm").GetValueOrDefault("password", "admin");

        return await RpcAsync("call", new
        {
            service = "object",
            method = "execute_kw",
            args = new object[] { _db, uid, password!, model, method, args, kwargs ?? new Dictionary<string, object>() }
        });
    }

    public async Task<int> CreateContactAsync(Dictionary<string, object> contactData)
    {
        var result = await ExecuteKwAsync("res.partner", "create", new[] { contactData });
        return result.GetInt32();
    }

    public async Task<int> SyncCustomerAsync(string id, string name, string email, bool isCompany = true)
    {
        var searchResult = await ExecuteKwAsync("res.partner", "search", new object[]
        {
            new object[] { new object[] { "ref", "=", id } }
        });

        var partnerData = new Dictionary<string, object>
        {
            { "name", name },
            { "email", email },
            { "ref", id },
            { "is_company", isCompany }
        };

        if (searchResult.ValueKind == JsonValueKind.Array && searchResult.GetArrayLength() > 0)
        {
            var existingId = searchResult[0].GetInt32();
            await ExecuteKwAsync("res.partner", "write", new object[] { new[] { existingId }, partnerData });
            return existingId;
        }

        return await CreateContactAsync(partnerData);
    }

    public async Task<bool> HealthAsync()
    {
        try
        {
            var result = await RpcAsync("call", new
            {
                service = "common",
                method = "version",
                args = Array.Empty<object>()
            });
            return result.ValueKind != JsonValueKind.Null;
        }
        catch
        {
            return false;
        }
    }
}

File: typescript/src/Crm/CrmManager.ts
/**
 * Zirve CRM Manager — Odoo Integration.
 *
 * Interacts with Odoo via JSON-RPC endpoint.
 */
export class CrmManager {
  private config: Record<string, any>;
  private url: string;
  private db: string;
  private uid: number | null = null;
  private sessionContext: Record<string, any> = {};
  
  // Track JSON-RPC id
  private rpcId = 0;

  constructor(config: Record<string, any>) {
    this.config = config;
    this.url = (config.url || 'http://odoo.zirve-infra.svc.cluster.local:8069').replace(/\/$/, '');
    this.db = config.database || 'zirve';
  }

  /**
   * Internal JSON-RPC executor.
   */
  private async rpc(method: string, params: Record<string, any>): Promise<any> {
    this.rpcId++;
    
    const response = await fetch(`${this.url}/jsonrpc`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method,
        params,
        id: this.rpcId
      })
    });

    if (!response.ok) {
      throw new Error(`Odoo RPC failed HTTP ${response.status}`);
    }

    const data = await response.json();
    if (data.error) {
      throw new Error(`Odoo RPC Error: ${data.error.message || JSON.stringify(data.error)}`);
    }

    return data.result;
  }

  /**
   * Authenticate and get a session.
   */
  private async authenticate(): Promise<number> {
    if (this.uid !== null) return this.uid;
    
    const user = this.config.username || 'admin';
    const password = this.config.password || 'admin';

    const result = await this.rpc('call', {
      service: 'common',
      method: 'authenticate',
      args: [this.db, user, password, {}]
    });

    if (!result) {
      throw new Error('Odoo Authentication failed.');
    } // result is the UID

    this.uid = result;
    return result;
  }

  /**
   * Execute an ORM method contextually.
   */
  private async executeKw(model: string, method: string, args: any[], kwargs: Record<string, any> = {}): Promise<any> {
    const uid = await this.authenticate();
    const password = this.config.password || 'admin';

    return await this.rpc('call', {
      service: 'object',
      method: 'execute_kw',
      args: [this.db, uid, password, model, method, args, kwargs]
    });
  }

  /**
   * Create a contact (partner) in Odoo.
   */
  public async createContact(contactData: Record<string, any>): Promise<number> {
    return await this.executeKw('res.partner', 'create', [contactData]);
  }

  /**
   * Create a CRM lead.
   */
  public async createLead(leadData: Record<string, any>): Promise<number> {
    return await this.executeKw('crm.lead', 'create', [leadData]);
  }

  /**
   * Create a helpdesk ticket.
   */
  public async createTicket(ticketData: Record<string, any>): Promise<number> {
    return await this.executeKw('helpdesk.ticket', 'create', [ticketData]);
  }

  /**
   * Sync a Zirve customer to Odoo. Returns partner ID.
   */
  public async syncCustomer(id: string, name: string, email: string, isCompany: boolean = true): Promise<number> {
    // Check if exists
    const existing = await this.executeKw('res.partner', 'search', [[['ref', '=', id]]]);
    
    const partnerData = {
      name,
      email,
      ref: id,
      is_company: isCompany
    };

    if (existing && existing.length > 0) {
      await this.executeKw('res.partner', 'write', [[existing[0]], partnerData]);
      return existing[0] as number;
    }

    return await this.createContact(partnerData);
  }

  /**
   * Basic Odoo health check.
   */
  public async health(): Promise<boolean> {
    try {
      const result = await this.rpc('call', {
        service: 'common',
        method: 'version',
        args: []
      });
      return !!result;
    } catch {
      return false;
    }
  }
}
